compiler/erlang_check.erl

```diff
@@ -279,6 +275,113 @@ check_module(File) ->
     end.
 
 %%------------------------------------------------------------------------------
+%% @doc Check for some known files and try to guess what build system is being
+%% used.
+%% @end
+%%------------------------------------------------------------------------------
+-spec guess_build_system(string()) -> {atom(), string()}.
+guess_build_system(Path) ->
+    BuildSystems = [
+                    {rebar, [
+                             "rebar.config",
+                             "rebar.config.script"
+                            ]
+                    },
+                    {make, [
+                            "Makefile"
+                           ]
+                    }
+                   ],
+    case guess_build_system(Path, BuildSystems) of
+        {_, []} -> unknown_build_system;
+        {BuildSystem, Files} -> {BuildSystem, Files}
+    end.
+
+guess_build_system(Path, [{BuildSystem, Files}|Rest]) ->
+    log("Try build system: ~p ~p~n", [BuildSystem]),
+    case find_files(Path, Files) of
+        [] -> guess_build_system(Path, Rest);
+        FoundFiles when is_list(FoundFiles) -> {BuildSystem, FoundFiles}
+    end.
+
+find_files(Path, Files) ->
+    find_files(Path, Files, Files).
+
+find_files("/", _Files, _AllFiles) ->
+    [];
+find_files(_Path, [], _AllFiles) ->
+    [];
+find_files(Path, [File|Rest], AllFiles) ->
+    AbsFile = absname(Path, File),
+    case filelib:is_regular(AbsFile) of
+        true -> log("Found build file: ~p~n", [AbsFile]), [AbsFile];
+        false -> [] ++ find_files(Path, Rest, AllFiles)
+    end ++ find_files(filename:dirname(Path), AllFiles, AllFiles).
+
+load_rebar_files([]) ->
+    error;
+load_rebar_files([ConfigFile|_Rest]) ->
+    ConfigResult = case filename:extension(ConfigFile) of
+                       ".script" -> file:script(ConfigFile);
+                       ".config" -> file:consult(ConfigFile)
+                   end,
+    case ConfigResult of
+        {ok, ConfigTerms} ->
+            log("rebar.config read: ~s~n", [ConfigFile]),
+            {ok, process_rebar_config(filename:dirname(ConfigFile), ConfigTerms)};
+        {error, Reason} ->
+            log_error("rebar.config consult failed:~n"),
+            file_error(ConfigFile, Reason),
+            error
+    end.
+
+%%------------------------------------------------------------------------------
+%% @doc Apply a rebar.config file.
+%%
+%% This function adds the directories in the rebar.config file to the code path
+%% and returns and compilation options to be used when compiling the file.
+%% @end
+%%------------------------------------------------------------------------------
+process_rebar_config(Path, Terms) ->
+    DepsDir = proplists:get_value(deps_dir, Terms, "deps"),
+    LibDirs = proplists:get_value(lib_dirs, Terms, []),
+    SubDirs = proplists:get_value(sub_dirs, Terms, []),
+    ErlOpts = proplists:get_value(erl_opts, Terms, []),
+    Includes = [{i, absname(Path, "include")},
```

Comment:
This `absname(Path, "include")` seems to be replacing the following lines in the old code:

```erlang
% ".." -> include
[{i, absname(Dir, "..")},
```

Was this replacement intentional or accidental?

The following comment explained why ".." is written instead of "include":

```erlang
% * include/ => ".." -> include. This is needed because files in src may
%                use `-include_lib("appname/include/f.hrl")`
```

To elaborate: imagine an application called myapp, which has a file called `myapp/include/h.hrl` and a `myapp/src/e.erl` file. The latter says this:

```erlang
-include_lib("appname/include/h.hrl")
```

The value of `absname(Dir, "..")` will be `/path/to/myapp/..`. When `include_lib` will be looking for `myapp/include/h.hrl`, it will check each include dir if it contains a `myapp*/include/h.hrl` file (until it finds one). The `/path/to/myapp/..` will contain such a file (`/path/to/myapp/../myapp*/include/h.hrl`), namely `/path/to/myapp/include/h.hrl`.

Without adding `absname(Dir, "..")`, compiling such files would fail. This can be easily reproduced:

```sh
$ git clone git@github.com:esl/elarm.git
Cloning into 'elarm'...
remote: Counting objects: 455, done.
remote: Total 455 (delta 0), reused 0 (delta 0), pack-reused 455
Receiving objects: 100% (455/455), 137.47 KiB | 0 bytes/s, done.
Resolving deltas: 100% (293/293), done.
Checking connectivity... done.

$ cd elarm

# vim-erlang-erlang using the proposed changes
$ .../vim-erlang-compiler/compiler/erlang_check.erl src/elarm.erl
/home/hcs/__/elarm/src/elarm.erl:68: can't find include lib "elarm/include/elarm.hrl"
/home/hcs/__/elarm/src/elarm.erl:143: type additional_information() undefined
/home/hcs/__/elarm/src/elarm.erl:172: type clear_reason() undefined
/home/hcs/__/elarm/src/elarm.erl:273: type sub_filter() undefined
/home/hcs/__/elarm/src/elarm.erl:336: type text() undefined
/home/hcs/__/elarm/src/elarm.erl:356: type alarm_src() undefined
/home/hcs/__/elarm/src/elarm.erl:356: type user_id() undefined
/home/hcs/__/elarm/src/elarm.erl:414: type alarm() undefined
/home/hcs/__/elarm/src/elarm.erl:414: type event_id() undefined
/home/hcs/__/elarm/src/elarm.erl:417: record alarm undefined
/home/hcs/__/elarm/src/elarm.erl:523: type alarm_config() undefined
/home/hcs/__/elarm/src/elarm.erl:523: type alarm_id() undefined

# vim-erlang-erlang at vim-erlang/master
$ .../vim-erlang-compiler/compiler/erlang_check.erl src/elarm.erl

$
```

I suggest that we don't modify the `process_rebar_config` at all in this commit. That way we will avoid introducing this problem, and I don't want to lose neither the comments nor the type specs (both of which are removed by the commit). The comments are there to help understanding; if they don't explain the situation well enough (e.g. why those ".." is needed), then we can improve them.

Also, probably the `load_makefiles` function should add the `..` directory as well (so that if a Makefile user has the directory layout that I described above, he will have have this problem).